2FA prevents publishing to npm

### Problem

Publishing packages to NPM fails if the user has two-factor authentication enabled on their NPM account:

```
lerna info publish Publishing packages to npm...
lerna ERR! publish Retrying failed publish: @material/animation
lerna ERR! publish Retrying failed publish: @material/animation
lerna ERR! publish Retrying failed publish: @material/animation
lerna ERR! publish Retrying failed publish: @material/animation
lerna ERR! publish Ran out of retries while publishing @material/animation Error: Command failed: npm publish --tag latest
lerna ERR! publish npm ERR! publish Failed PUT 401
lerna ERR! publish npm ERR! code E401
lerna ERR! publish npm ERR! 404 You must provide a one-time pass. : @material/animation
```

The "official" workaround is to set a `NPM_CONFIG_OTP` env var, but that only works if you have a small number of packages.

Publishing ~10 or more packages takes longer than the 30 second OTP timeout, which causes the first few packages to be uploaded successfully whereas the remainder fail.

At that point, you have to manually check every package on npmjs.com and see if it was updated or not, and run `npm publish` on any packages that failed.

See Lerna issues [#1137](https://github.com/lerna/lerna/issues/1137) and [#1091](https://github.com/lerna/lerna/issues/1091) for more information.

### Solution

We should:

* Add instructions on how to temporarily disable and subsequently re-enable 2FA in \
    [`release-process.md`](https://github.com/material-components/material-components-web/blob/5b1348cee8302f12d552e8612d911a202be49ad5/docs/open_source/release-process.md)
* Check for 2FA in [`pre-release.sh`](https://github.com/material-components/material-components-web/blob/5b1348cee8302f12d552e8612d911a202be49ad5/scripts/pre-release.sh#L33-L39)

There's also a small typo in `release-process.md`: `npm deploy` should be `npm run deploy`.

--- a/docs/open_source/release-process.md
+++ b/docs/open_source/release-process.md
@@ -73,6 +73,16 @@ Run the pre-release script:
 This will ensure you can publish/tag, build all release files, and ensure all tests
 pass prior to releasing (lerna will update versions for us in the next step).
 
+## Disable 2FA
+
+If you have two-factor authentication enabled on your NPM account (you should), you'll need to temporarily disable it:
+
+```bash
+npm profile disable-2fa
+```
+
+See Lerna issues [#1137](https://github.com/lerna/lerna/issues/1137) and [#1091](https://github.com/lerna/lerna/issues/1091) for more information.
+
 ## Publish to npm
 
 ### For Pre-releases
@@ -118,6 +128,16 @@ When lerna prompts for version, choose Patch.
 
 Be sure to include the command-line flag; `--skip-git` avoids immediately applying a git tag, which we will do later
 after updating the changelog.
+
+## Enable 2FA
+
+If you temporarily disabled two-factor authentication on your NPM account, you'll need to re-enable it:
+
+```bash
+npm profile enable-2fa auth-and-writes
+```
+
+See Lerna issues [#1137](https://github.com/lerna/lerna/issues/1137) and [#1091](https://github.com/lerna/lerna/issues/1091) for more information.
 
 ## Commit Version Bumps
 
@@ -199,7 +219,7 @@ git push
 
 1. Update the `material-components-web` dependency in the catalog's `package.json` to the new patch version
 1. Run `npm start` and glance through the catalog pages to make sure everything looks normal
-1. Send a PR for the dependency update, then run `npm deploy` once it's merged to master
+1. Send a PR for the dependency update, then run `npm run deploy` once it's merged to master
 
 ### For Minor Releases
 

--- a/scripts/pre-release.sh
+++ b/scripts/pre-release.sh
@@ -28,26 +28,41 @@ function log() {
   echo '\033[36m[pre-release]\033[0m' "$@"
 }
 
+function fail() {
+  echo '\033[31mFAILURE:\033[0m' "$@"
+}
+
 log "Running pre-flight sanity checks..."
 
 log "Checking that you can publish to npm..."
 NPM_USER=$(npm whoami)
 if ! npm team ls material:developers | grep -q $NPM_USER; then
-  echo "FAILURE: You are not (yet?) part of the material:developers org. Please get in touch" \
+  fail "You are not (yet?) part of the material:developers org. Please get in touch" \
        "with the MDC Web core team to rectify this"
+  exit 1
+fi
+
+log "Checking that npm two-factor authentication is disabled..."
+NPM_TFA=$(npm profile get --parseable | grep '^tfa' | awk '{ print $2 }')
+if [[ "$NPM_TFA" != 'disabled' ]]; then
+  fail "Two-factor authentication (2FA) is enabled on your NPM account, which prevents publishing."
+  echo "To temporarily disable it, edit your Profile Settings on https://www.npmjs.com/, or run the following command:"
+  echo "    npm profile disable-2fa"
+  echo "After publishing, re-enable 2FA on https://www.npmjs.com/, or run the following command:"
+  echo "    npm profile enable-2fa auth-and-writes"
   exit 1
 fi
 
 log "Checking that you can access GitHub via SSH..."
 if ! ssh -T git@github.com 2>&1 | grep -q "You've successfully authenticated"; then
-  echo "FAILURE: It does not look like you can access github. Please ensure that the command" \
+  fail "It does not look like you can access github. Please ensure that the command" \
        "ssh -T git@github.com works for you"
   exit 1
 fi
 
 log "Checking that you can deploy the MDC Web demo site..."
 if ! gcloud config get-value project 2>/dev/null | grep -q material-components-web; then
-  echo "FAILURE: Your gcloud project is not configured for MDC Web. Please run gcloud config set" \
+  fail "Your gcloud project is not configured for MDC Web. Please run gcloud config set" \
        "project material-components-web and ensure it exits successfully"
   exit 1
 fi

